src/resolve/resolved-mdns-browse-services.h

```diff
@@ -0,0 +1,78 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#pragma once
+
+typedef struct mDnsServiceSubscriber mDnsServiceSubscriber;
+
+#include "resolved-manager.h"
+#include "varlink.h"
+
+typedef struct mDnsServices mDnsServices;
+
+typedef enum mDnsRecordTTLState mDnsRecordTTLState;
+
+enum mDnsRecordTTLState {
+        MDNS_TTL_80_PERCENT,
+        MDNS_TTL_85_PERCENT,
+        MDNS_TTL_90_PERCENT,
+        MDNS_TTL_95_PERCENT,
+        MDNS_TTL_100_PERCENT
+};
+
+struct mDnsServices {
+        unsigned n_ref;
+        mDnsServiceSubscriber *ss;
+        sd_event_source *schedule_event;
+        DnsResourceRecord *rr;
+        int family;
+        usec_t until;
+        mDnsRecordTTLState rr_ttl_state;
+        DnsQuery *query;
+        LIST_FIELDS(mDnsServices, mdns_services);
+};
+
+struct mDnsServiceSubscriber {
+        unsigned n_ref;
+        Manager *m;
+        Varlink *link;
+        DnsQuestion *question_idna;
+        DnsQuestion *question_utf8;
+        uint64_t flags;
+        sd_event_source *schedule_event;
+        usec_t delay;
+        DnsResourceKey *key;
+        int ifindex;
+        uint64_t token;
+        LIST_HEAD(mDnsServices, mdns_services);
+};
+
+mDnsServiceSubscriber *mdns_service_subscriber_free(mDnsServiceSubscriber *ss);
```

Comment:
so the concept is a bit of a misnomer, as service browsing is not specific to mdns really, it can also be done with classic dns, though it's not commonly done.

maybe call this DnsServiceBrowser or so

Reply:
hmm, shouldn't this be folded into DnsQuery btw, it sounds so similar?

Reply:
Do you mean that this entire browse implemntation could also be used for DNS browsing?
Or that the mDnsServiceSubscriber struct could be generalized to be reused for DNS browsing?
At least currently, the implemntation (specifically the timing of queries) is according to section 5.2 of RFC6762, which is for mDNS.
Would we then use this same logic for DNS browsing?

Also, about folding it into DnsQuery, which part do you mean exactly?
Do you mean move this browse implemntation to extend DnsQuery?

Reply:
well, conceptually dns-sd allows to "browse" also on classic unicast dns. it would not really be long-running then, but short-running only, but i think the browse concept we expose should abstract this fact.

i.e. if you create a browser object on a unicast dns domain, then we'll immediately return the answers we'll get and be done, as opposed to mdns where we'll generate further events as the browser remains open.

(i mean, we could eventually even implement DNS-LLQ, i.e. RFC 8764 to implement browsing fully on unicast dns, but I think that's low-prio as it doesn't appear to be commonly deployed. But even if we don't care about this right now I think the IPC iface we provide should leave the door open to that)

Reply:
Changed to a more generic DnsServiceBrowser.
However, we did add a protocol check in dns_subscribe_browse_service to support only mdns at the moment since the timing of queries etc are in accordance with RFC6762, which covers mdns.
This could be extended to support other protocols as well.